File: 2022/go/4_camp_cleanup/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

type Tuple struct {
	left  int
	right int
}

func (t Tuple) Size() int {
	return t.right - t.left
}

func main() {
	f, err := os.Open("data.txt")
	if err != nil {
		panic(err)
	}

	var count int

	sc := bufio.NewScanner(f)
	for sc.Scan() {
		line := sc.Text()

		// We need to identify the bigger and smaller intervals (bigger as in "covers more numbers")
		elves := strings.Split(line, ",")

		var intervals [2]Tuple
		for i, elf := range elves {
			interval := strings.Split(elf, "-")
			left, right := interval[0], interval[1]

			l, err := strconv.Atoi(left)
			if err != nil {
				panic(err)
			}

			r, err := strconv.Atoi(right)
			if err != nil {
				panic(err)
			}

			intervals[i] = Tuple{left: l, right: r}
		}

		var bigger, smaller Tuple = intervals[0], intervals[1]
		if intervals[0].Size() < intervals[1].Size() {
			bigger = intervals[1]
			smaller = intervals[0]
		}

		// That took a lot of lines oof.

		// Next up we check the borders and we're done.
		if bigger.left <= smaller.left && smaller.right <= bigger.right {
			count++
		}
	}

	fmt.Println(count)
}

File: 2022/rust/4_camp_cleanup/src/main.rs
use std::fs::read_to_string;
use std::str::FromStr;
use std::num::ParseIntError;

fn main() {
    let content = read_to_string("data.txt").unwrap();

    let mut count = 0;
    for (_, line) in content.lines().enumerate() {
        let mut pairs: Vec<Pair> = line.split(",").map(|x| x.parse().unwrap()).collect();

        pairs.sort_by_key(|x| -x.size());

        let (bigger, smaller) = (pairs.get(0).unwrap(), pairs.get(1).unwrap());

        if bigger.left <= smaller.left && smaller.right <= bigger.right {
            count += 1;
        }
    }

    println!("{}", count);
}

#[derive(Debug)]
struct Pair {
    left: i32,
    right: i32,
}

impl Pair {
    fn size(&self) -> i32 {
        self.right - self.left
    }
}

impl FromStr for Pair {
    type Err = ParseIntError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let (left, right) = s.split_once("-").unwrap();
        let (l, r): (i32, i32) = (left.parse().unwrap(), right.parse().unwrap());

        Ok(Pair { left: l, right: r})
    }
}
